### PR TITLE
Adding onEnd callback to implicit animated widgets

### PR DESCRIPTION
**Pull request by RobinJespersen**
## Description
There is now an onEnd VoidCallback that is called at the end of the implicit animation.

It is added to the following widgets:
-  AnimatedContainer
-  AnimatedPadding
-  AnimatedAlign
-  AnimatedPositioned
-  AnimatedPositionedDirectional
-  AnimatedOpacity
-  AnimatedDefaultTextStyle
-  AnimatedPhysicalModel
-  AnimatedTheme

Furthermore: moved onEnd implementation from TweenAnimationBuilder to ImplicitlyAnimatedWidget

## Related Issues
#34799

## Tests

I added the following tests:
A test that ensures the callback is called.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by RobinJespersen** [comment]
@googlebot I fixed it.

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 1]
```diff
@@ -47,4 +74,133 @@ void main() {
...
+
+  void onChanged(bool v) {
+    setState(() {
+      toggle = v;
+    });
+  }
+
+  Widget getAnimatedWidget() {
+    const Widget child = Placeholder();
+    final Duration duration = Duration(milliseconds: 10);
```
Comment (Piinks) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: const

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 2]
```diff
@@ -47,4 +74,133 @@ void main() {
...
+          onEnd: widget.callback,
+          opacity: toggle ? 0.1 : 0.9,
+        );
+      case ImplicitAnimatedWidgetType.AnimatedDefaultTextStyle:
+        return AnimatedDefaultTextStyle(
+            child: child,
+            duration: duration,
+            onEnd: widget.callback,
+            style: toggle
+                ? TextStyle(fontStyle: FontStyle.italic)
```
Comment (Piinks) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: const

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 3]
```diff
@@ -47,4 +74,133 @@ void main() {
...
+                : TextStyle(fontStyle: FontStyle.normal));
```
Comment (Piinks) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: const

**Review by Piinks — COMMENTED** [review]
Hi @RobinJespersen !
Welcome!  🎉  And thank you for the contribution. 
I believe this would also need to be implemented in `AnimatedTheme`?
The documentation for [`ImplicitlyAnimatedWidget`](https://master-api.flutter.dev/flutter/widgets/ImplicitlyAnimatedWidget-class.html) lists all of its implementers, and I think this change covers all but that one. :) 
It also seems that there are some analyzer issues from our pre-submit tests regarding const constructors. I've noted them below.

**Comment by RobinJespersen** [comment]
@Piinks thx for the hints, they helped a lot :)

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 4]
```diff
@@ -47,4 +76,154 @@ void main() {
...
+    for (ImplicitAnimatedWidgetType widgetType
+        in ImplicitAnimatedWidgetType.values) {
+      await tester.pumpWidget(wrap(
+          child: TestAnimatedWidget(mockOnEndFunction, switchKey, widgetType)));
+
+      final Finder widgetFinder = find.byKey(switchKey);
+
+      await tester.tap(widgetFinder);
+
+      await tester.pumpAndSettle();
```
Comment (chunhtai) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: I would use a for loop to pump frame one by one and make sure onEnd only called at the very end
```dart
// Assume animation duration is 1 second.
await tester.pump();
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, ++i);
```

File: packages/flutter/lib/src/widgets/tween_animation_builder.dart [hunk 5]
```diff
@@ -29,8 +29,8 @@ import 'value_listenable_builder.dart';
 /// current animation value. The [builder] is called throughout the animation
 /// for every animation value until [Tween.end] is reached.
 ///
-/// A provided [onEnd] callback is called whenever an animation completes.
-/// Registering an [onEnd] callback my be useful to trigger an action (like
+/// A provided [ImplicitlyAnimatedWidget.onEnd] callback is called whenever an animation completes.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/tween_animation_builder.dart]: I think we can move this comment to constructor and keep [onEnd]

**Comment by goderbauer** [comment]
@RobinJespersen Do you have plans to come back to this PR and address the comments?

**Comment by RobinJespersen** [comment]
Yes, just need to find time for it.Am 04.09.2019 23:43 schrieb Michael Goderbauer <notifications@github.com>:@RobinJespersen Do you have plans to come back to this PR and address the comments?

File: packages/flutter/lib/src/widgets/implicit_animations.dart [hunk 6]
```diff
@@ -354,6 +358,12 @@ abstract class ImplicitlyAnimatedWidgetState<T extends ImplicitlyAnimatedWidget>
       debugLabel: kDebugMode ? '${widget.toStringShort()}' : null,
       vsync: this,
     );
+    _controller.addStatusListener((AnimationStatus status) {
+      if (status == AnimationStatus.completed) {
```
[the remark this answers is not in the document]
Comment (RobinJespersen) [reply packages/flutter/lib/src/widgets/implicit_animations.dart]: done

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 7]
```diff
@@ -2,10 +2,39 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:flutter/material.dart';
 import 'package:flutter_test/flutter_test.dart';
 import 'package:flutter/widgets.dart';
 
+enum ImplicitAnimatedWidgetType {
```
[the remark this answers is not in the document]
Comment (RobinJespersen) [reply packages/flutter/test/widgets/implicit_animations_test.dart]: done

File: packages/flutter/test/widgets/implicit_animations_test.dart (hunk 4 above)
Comment (RobinJespersen) [reply packages/flutter/test/widgets/implicit_animations_test.dart]: I guess this should be sufficient:  
```
    await tester.pump();
    expect(mockOnEndFunction.called, 0);
    await tester.pump(const Duration(milliseconds:1000));
    expect(mockOnEndFunction.called, 0);
    await tester.pump(const Duration(milliseconds:1));
    expect(mockOnEndFunction.called, 1);
```

File: packages/flutter/lib/src/widgets/tween_animation_builder.dart (hunk 5 above)
Comment (RobinJespersen) [reply packages/flutter/lib/src/widgets/tween_animation_builder.dart]: I am not sure what you mean...

Comment (goderbauer) [inline packages/flutter/lib/src/widgets/tween_animation_builder.dart]: DartDoc can link to this by just doing the following:

```suggestion
/// A provided [onEnd] callback is called whenever an animation completes.
```

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 8]
```diff
@@ -47,4 +63,355 @@ void main() {
...
+  @override
+  Widget getAnimatedWidget() {
+    return AnimatedTheme(
+      child: child,
+      data: toggle ? ThemeData.dark() : ThemeData.light(),
+      duration: duration,
+      onEnd: widget.callback,
+    );
+  }
+}
```
Comment (goderbauer) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: nit: end with blank line.

File: packages/flutter/test/widgets/implicit_animations_test.dart [hunk 9]
```diff
@@ -47,4 +63,355 @@ void main() {
...
+        child: TestAnimatedWidget(callback: mockOnEndFunction, switchKey: switchKey, state: _TestAnimatedContainerWidgetState(),)
+    ));
+
+    final Finder widgetFinder = find.byKey(switchKey);
+
+    await tester.tap(widgetFinder);
+
+    await tester.pump();
+    expect(mockOnEndFunction.called, 0);
+    await tester.pump(const Duration(milliseconds:1000));
```
Comment (goderbauer) [inline packages/flutter/test/widgets/implicit_animations_test.dart]: can this duration be a const defined somewhere that gets reused everywhere?

**Review by goderbauer — COMMENTED** [review]
This looks good. Just a couple of nits.

File: packages/flutter/lib/src/widgets/tween_animation_builder.dart (hunk 5 above)
Comment (RobinJespersen) [reply packages/flutter/lib/src/widgets/tween_animation_builder.dart]: thx

**Review by Piinks — APPROVED** [review]
LGTM! Thank you again for the contribution!

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by esDotDev** [comment]
@RobinJespersen @goderbauer Would be be a big lift to get an onUpdate callback implemented? 

Please see my rationale here: 
https://github.com/flutter/flutter/issues/34799#issuecomment-538654842

The current implementation skews too hard towards inheritence imo, and is not as friendly to composition as it could be.

**Pull request merged**